system/Files/FileCollection.php

```diff
@@ -0,0 +1,386 @@
+<?php
+
+/**
+ * This file is part of the CodeIgniter 4 framework.
+ *
+ * (c) CodeIgniter Foundation <admin@codeigniter.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace CodeIgniter\Files;
+
+use CodeIgniter\Files\Exceptions\FileException;
+use CodeIgniter\Files\Exceptions\FileNotFoundException;
+use Countable;
+use Generator;
+use InvalidArgumentException;
+use IteratorAggregate;
+use Traversable;
+
+/**
+ * File Collection Class
+ *
+ * Representation for a group of files, with utilities for locating,
+ * filtering, and ordering them.
+ */
+class FileCollection implements Countable, IteratorAggregate
+{
+	/**
+	 * The current list of file paths.
+	 *
+	 * @var string[]
+	 */
+	protected $files = [];
+
+	//--------------------------------------------------------------------
+
+	/**
+	 * Resolves a full path and verifies it is an actual directory.
+	 *
+	 * @param string $directory
+	 *
+	 * @return string
+	 *
+	 * @throws FileException
+	 */
+	protected static function resolveDirectory(string $directory): string
+	{
+		if (! is_dir($directory = set_realpath($directory)))
+		{
+			$caller = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2)[1];
+			throw FileException::forExpectedDirectory($caller['function']);
+		}
+
+		return $directory;
+	}
+
+	/**
+	 * Resolves a full path and verifies it is an actual file.
+	 *
+	 * @param string $file
+	 *
+	 * @return string
+	 *
+	 * @throws FileException
+	 */
+	protected static function resolveFile(string $file): string
+	{
+		if (! is_file($file = set_realpath($file)))
+		{
+			$caller = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2)[1];
+			throw FileException::forExpectedFile($caller['function']);
+		}
+
+		return $file;
+	}
+
+	/**
+	 * Removes files that are not part of the given directory (recursive).
+	 *
+	 * @param string[] $files
+	 * @param string   $directory
+	 *
+	 * @return string[]
+	 */
+	protected static function filterFiles(array $files, string $directory): array
+	{
+		$directory = self::resolveDirectory($directory);
```

Comment:
since the class is not final, if the `resolveDirectory()` is `protected`, should it be called via `static::` ?

Reply:
Hmm good point. This moved from Publisher where it was private, but since Publisher extends FileCollection and still needs that method I loosed it to protected. Maybe these static support functions should all be final? Or use the static call like you said?

If anyone has thoughts on the scoping in this class I'm open to suggestions.

Reply:
To be honest, I've never really been a fan of final classes/methods in a framework since that limits the end-users ability to extend the files. Or maybe I've never been confident enough that my ability to see the future was good enough lol. Always seemed like a better feature for applications than a framework. 

For a class like this, I'm not against it, though, since it functions as a security layer of sorts. Though people could maliciously insert a mock version of the class with new features? NVM that's only available during testing I believe. Either way, we can't stop everything. 

tl;dr; I honestly have no preference for whether this is a final class or not, but am not opposed if that's what others prefer.

Reply:
> maybe I've never been confident enough that my ability to see the future was good enough lol

This is actually why I'm a fan of `final` in many cases! We can be certain that nobody has extended the class/method so it is safe to make more changes than usual, and there is no penalty for removing `final` down the road.

In this case I decided to make the static helper methods `final` since they are essentially the equivalent of a new helper file - should this become an issue we can remove the `final` and call them with `static::` later.